test/Npgsql.Tests/Replication/PgOutputReplicationTests.cs

```diff
@@ -0,0 +1,537 @@
+                    await using var c = await OpenConnectionAsync();
+                    TestUtil.MinimumPgVersion(c, "10.0", "The Logical Replication Protocol (via pgoutput plugin) was introduced in PostgreSQL 10");
+                    await using var rc = await OpenReplicationConnectionAsync();
+                    var options = await rc.CreateReplicationSlot(slotName);
+
+                    using var cmd =
+                        new NpgsqlCommand($"SELECT * FROM pg_replication_slots WHERE slot_name = '{options.SlotName}'",
+                            c);
+                    await using var reader = await cmd.ExecuteReaderAsync();
+
+                    Assert.That(reader.Read, Is.True);
+                    Assert.That(reader.GetFieldValue<string>(reader.GetOrdinal("slot_type")), Is.EqualTo("logical"));
+                    Assert.That(reader.GetFieldValue<string>(reader.GetOrdinal("plugin")), Is.EqualTo("pgoutput"));
+                    Assert.That(reader.Read, Is.False);
+                });
+
+        [Test(Description = "Tests whether INSERT commands get replicated as Logical Replication Protocol Messages")]
+        public Task StartReplicationReplicatesInsert()
+            => SafeReplicationTest(nameof(StartReplicationReplicatesInsert) + "_pgoutput",
+                async (slotName, tableName, publicationName) =>
+                {
+                    await using var c = await OpenConnectionAsync();
+                    TestUtil.MinimumPgVersion(c, "10.0", "The Logical Replication Protocol (via pgoutput plugin) was introduced in PostgreSQL 10");
+                    var messages = new ConcurrentQueue<LogicalReplicationProtocolMessage>();
+                    await c.ExecuteNonQueryAsync(@$"
+CREATE TABLE {tableName} (id INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY, name TEXT NOT NULL);
+CREATE PUBLICATION {publicationName} FOR TABLE {tableName};
+");
+                    var rc = await OpenReplicationConnectionAsync();
+                    var slot = await rc.CreateReplicationSlot(slotName);
+                    await c.ExecuteNonQueryAsync($"INSERT INTO {tableName} (name) VALUES ('val1'), ('val2')");
+
+                    using var streamingCts = new CancellationTokenSource();
+                    var replicationTask = Task.Run(async () =>
+                    {
+                        await foreach (var msg in (await rc.StartReplication(slot, new NpgsqlPgOutputPluginOptions(publicationName), cancellationToken: streamingCts.Token)).WithCancellation(streamingCts.Token))
+                            messages.Enqueue(msg);
+                    }, CancellationToken.None);
+
+                    // Begin Transaction
+                    var message = await DequeueMessage(messages);
+                    Assert.That(message, Is.TypeOf<BeginMessage>());
+
+                    // Relation
+                    message = await DequeueMessage(messages);
+                    Assert.That(message, Is.TypeOf<RelationMessage>());
```

Comment:
nit: I'd just skip this, the next line would throw anyway (IIRC xunit has a nice thing which asserts the type and returns after casting, but nunit isn't as good on that stuff)

Reply:
Yeah, that's true if you only care about the test result.
An `AssertionFailedException` gives you a somewhat better problem description than an `InvalidCastException` if you don't look at the source code, but that's probably only cosmetics.

Reply:
Up to you. In general I most care about the test failing if things go wrong :)

Reply:
FYI I wrapped this in a small utility function in https://github.com/Brar/npgsql/pull/4